### PR TITLE
`Programming exercises`: Improve markdown editor layout handling

### PR DESCRIPTION
**Pull request by pzdr7**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

The markdown editor currently uses absolute positioning to place the resize element. This works for the use case of editing a programming exercise, but is bound to break for more complex use cases such as messaging.

### Description

- Includes the resize handle in the layout of the markdown editor.
- Removes the absolute positioning and the previous resize handling and replaces it with a different layout approach
- Various styling adjustments

### Steps for Testing

**Known issues**:

- The "Save" button can overlap with the action bar. This will be fixed in a follow-up

Prerequisites:
- 1 Instructor
- 1 Programming Exercise

**Edit in Editor**

1. Go to course management > Your course > Exercises > Your programming exercise and click the title of your exercise.
2. Click "Edit in Editor"
3. Play around with the layout of the markdown editor:
  - Switch between preview and editor.
  - Resize the markdown editor vertically and horizontally. Ensure that you can always see all of the markdown editor
  - Click the fullscreen button. Make sure the editor actually takes up the entire screen.

**Edit**

5. Go back and click "Edit"
6. Go to the problem statement and play around with the vertical resize. Verify that the min/max height still work and that no elements disappear.

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| markdown-editor-monaco.component.ts | 96.46% | ✅ |

### Screenshots

Only small enhancements

## Summary by CodeRabbit

- **New Features**
  - Added improved dynamic resizing functionality to the markdown editor.
  - Introduced constants for better height management in the editor.

- **Enhancements**
  - Refined HTML layout for the markdown editor to enhance usability.
  - Updated CSS styles to improve visual elements and user interaction.
  - Improved logic for element positioning during resizing for a smoother user experience.

- **Tests**
  - Added new test cases for fullscreen action in the markdown editor.

### CODE REVIEW

File: src/main/webapp/app/shared/markdown-editor/monaco/markdown-editor-monaco.component.html [hunk 1]
```diff
@@ -1,134 +1,130 @@
...
-                                    </a>
-                                </div>
-                            </label>
-                        </div>
-                    }
-                    @if (enableResize) {
-                        <!-- This placeholder will hold the resize handle. -->
-                        <span #resizablePlaceholder></span>
-                    }
-                </div>
```
Comment (pzdr7) [inline src/main/webapp/app/shared/markdown-editor/monaco/markdown-editor-monaco.component.html:44]: The diff is a little hard to read here. Basically, I removed this part and wrapped the #wrapper div in a #fullElement div. I then added the #resizePlaceholder as a separate div below the #wrapper but inside the #fullElement

File: src/main/webapp/app/shared/markdown-editor/monaco/markdown-editor-monaco.component.html [hunk 2]
```diff
@@ -1,134 +1,130 @@
...
-                                }
+<div #fullElement class="h-100">
+    <div
+        #wrapper
+        class="markdown-editor-wrapper"
+        [style.overflow]="inPreviewMode ? 'auto' : 'hidden'"
+        [style.height.px]="inPreviewMode ? undefined : targetWrapperHeight"
+        [style.cursor]="isResizing ? 'ns-resize' : undefined"
+    >
+        <nav ngbNav #actionPalette #nav="ngbNav" class="nav-tabs" [destroyOnHide]="false" (navChange)="onNavChanged($event)">
```
Comment (pzdr7) [inline src/main/webapp/app/shared/markdown-editor/monaco/markdown-editor-monaco.component.html:9]: I changed this part

File: src/main/webapp/app/shared/markdown-editor/monaco/markdown-editor-monaco.component.html [hunk 3]
```diff
@@ -1,134 +1,130 @@
...
+                cdkDragLockAxis="y"
+                [cdkDragConstrainPosition]="constrainDragPositionFn"
+                (cdkDragStarted)="isResizing = true"
+                (cdkDragMoved)="onResizeMoved($event)"
+                (cdkDragEnded)="isResizing = false"
+                [icon]="faGripLines"
+                [style.cursor]="'ns-resize'"
+                class="rg-bottom md-resize-icon"
+            />
+        </div>
```
Comment (pzdr7) [inline src/main/webapp/app/shared/markdown-editor/monaco/markdown-editor-monaco.component.html:124]: This part is the new resize placeholder

File: src/main/webapp/app/shared/markdown-editor/monaco/markdown-editor-monaco.component.ts [hunk 4]
```diff
@@ -37,6 +37,9 @@ interface MarkdownActionsByGroup {
     meta: MonacoEditorAction[];
 }
 
+const EXTERNAL_HEIGHT = 'external';
+const BORDER_WIDTH_OFFSET = 3;
```
Comment (pzdr7) [inline src/main/webapp/app/shared/markdown-editor/monaco/markdown-editor-monaco.component.ts:41]: Why 3? -> The editor will often cover the container border if I set it to the exact value (2). This preserves the border at the cost of it sometimes being a little too thick (though the editor scrollbar is probably in the way already)

Comment (JohannesStoehr) [reply src/main/webapp/app/shared/markdown-editor/monaco/markdown-editor-monaco.component.ts:41]: Can you also write this as a comment in the file?

Comment (pzdr7) [reply src/main/webapp/app/shared/markdown-editor/monaco/markdown-editor-monaco.component.ts:41]: Yes

**Review by JohannesStoehr — APPROVED** [review]
Code lgtm

**Review by Jan-Thurner — APPROVED** [review]
Code LGTM

**Review by coolchock — APPROVED** [review]
tested on ts4, looks good to me! 👍

**Review by egekurt123 — APPROVED** [review]
Works as expected on TS1

**Pull request merged**